### PR TITLE
Add support for undeleteable cells

### PR DESCRIPTION
**Pull request by jjaraalm**
Implements #1310. Checks each cell for the presence of the `deletable` metadata and handles it appropriately. Also unifies cut, copy, and delete in order to reduce code reduplication. Still need to update testing, but would like advice on how to do that.

### CODE REVIEW

File: src/notebook/notebook/actions.ts [hunk 1]
```diff
@@ -178,32 +178,38 @@ namespace NotebookActions {
...
-    // The model will add a new code cell if there are no
-    // remaining cells.
-    model.cells.endCompoundOperation();
-
-    // Select the cell *after* the last selected.
-    // Note: The activeCellIndex is clamped to the available cells,
-    // so if the last cell is deleted the previous cell will be activated.
-    index -= toDelete.length - 1;
-    widget.activeCellIndex = index;
+    // If cells are not deletable, we may not have anything to delete
```
Comment (blink1073) [inline src/notebook/notebook/actions.ts]: Please end all comments in a period.

File: src/notebook/notebook/actions.ts [hunk 2]
```diff
@@ -178,32 +178,38 @@ namespace NotebookActions {
...
+    if(toDelete.length > 0){
```
Comment (blink1073) [inline src/notebook/notebook/actions.ts]: spaces around the parens

File: src/notebook/notebook/actions.ts [hunk 3]
```diff
@@ -609,6 +615,35 @@ namespace NotebookActions {
   }
 
   /**
+   * Copy or cut the selected cell data to a clipboard.
+   *
+   * @param widget - The target notebook widget.
+   *
+   * @param clipboard - The clipboard object.
+   *
+   * @param cut - Whether to copy or cut
```
Comment (blink1073) [inline src/notebook/notebook/actions.ts]: missing period

File: src/notebook/notebook/actions.ts [hunk 4]
```diff
@@ -609,6 +615,35 @@ namespace NotebookActions {
...
+     widget.mode = 'command';
+     clipboard.clear();
+     let data: nbformat.IBaseCell[] = [];
+     each(widget.widgets, child => {
+       if (widget.isSelected(child)) {
+         data.push(child.model.toJSON());
+       }
+     });
+     clipboard.setData(JUPYTER_CELL_MIME, data);
+     if(cut){
```
Comment (blink1073) [inline src/notebook/notebook/actions.ts]: spaces around parens

File: src/notebook/notebook/actions.ts [hunk 5]
```diff
@@ -617,19 +652,7 @@ namespace NotebookActions {
...
-    clipboard.clear();
-    let data: nbformat.IBaseCell[] = [];
-    each(widget.widgets, child => {
-      if (widget.isSelected(child)) {
-        data.push(child.model.toJSON());
-      }
-    });
-    clipboard.setData(JUPYTER_CELL_MIME, data);
-    widget.deselectAll();
+    copyOrCut(widget,clipboard,false);
```
Comment (blink1073) [inline src/notebook/notebook/actions.ts]: spaces after commas

File: src/notebook/notebook/actions.ts [hunk 6]
```diff
@@ -645,37 +668,7 @@ namespace NotebookActions {
...
-
-    // If there are no cells, add a code cell.
-    if (!model.cells.length) {
-      let cell = model.factory.createCodeCell();
-      model.cells.pushBack(cell);
-    }
-    model.cells.endCompoundOperation();
-
-    clipboard.setData(JUPYTER_CELL_MIME, data);
+    copyOrCut(widget,clipboard,true);
```
Comment (blink1073) [inline src/notebook/notebook/actions.ts]: spaces after commas

**Comment by jjaraalm** [comment]
Thanks for the comments, everything should be cleaned up now..

**Comment by blink1073** [comment]
Thanks again!

**Pull request merged**